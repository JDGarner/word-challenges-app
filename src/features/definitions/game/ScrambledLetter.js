import React, { useEffect } from "react";
import { View, Animated } from "react-native";
import styled from "styled-components";
import { BorderedButton, MediumText } from "../../../components";
import { TEXT_TOP_PADDING } from "../../../components/text/Text";
import { animateScrambledLetter } from "../definitions-utils";
import SoundManager from "../../sound/SoundManager";
import { getSizingForOptions } from "../../../utils/sizing-utils";

const LETTER_SIZE = getSizingForOptions(44, 45, 50);
const MARGIN_SIZE = getSizingForOptions(5, 6, 6);

const EmptyLetterPlaceHolder = styled(View)`
  margin-vertical: ${MARGIN_SIZE};
  margin-horizontal: ${MARGIN_SIZE};
  height: ${LETTER_SIZE};
  width: ${LETTER_SIZE};
`;

const LetterButton = styled(BorderedButton)`
  justify-content: center;
  align-items: center;
  margin-vertical: ${MARGIN_SIZE};
  margin-horizontal: ${MARGIN_SIZE};
  height: ${LETTER_SIZE};
  width: ${LETTER_SIZE};
  padding-top: ${TEXT_TOP_PADDING};
`;

const ScrambledLetter = ({ letter, showing, onPressLetter, disabled, scaleValue, isShuffling }) => {
  useEffect(() => {
    if (!isShuffling) {
      animateScrambledLetter(scaleValue, showing);
    }
  }, [showing]);

  const onPressLetterButton = () => {
    SoundManager.getInstance().playAddLetterSound();
    onPressLetter();
  };

  return (
    <Animated.View style={{ transform: [{ scale: scaleValue }], opacity: scaleValue }}>
      {showing === false ? (
        <EmptyLetterPlaceHolder />
      ) : (
        <LetterButton onPress={onPressLetterButton} reduceScaleFactor={0.8} disabled={disabled}>
          <MediumText>{letter}</MediumText>
        </LetterButton>
      )}
    </Animated.View>
  );
};

export default ScrambledLetter;
