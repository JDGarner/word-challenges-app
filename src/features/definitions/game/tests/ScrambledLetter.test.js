import "react-native";
import React from "react";
import { renderWithTheme } from "../../../../testing";
import ScrambledLetter from "../ScrambledLetter";
import * as ReactNative from "react-native";

describe("Scrambled Letter", () => {
  it("should match snapshot", () => {
    const mockFn = jest.fn();

    const rendered = renderWithTheme(
      <ScrambledLetter
        showing
        letter="S"
        isShuffling={false}
        disabled={false}
        scaleValue={new ReactNative.Animated.Value(1)}
        onPressLetter={mockFn}
      />,
    ).toJSON();

    expect(rendered).toMatchSnapshot();
  });

  it("should show empty place holder when showing is false", () => {
    const mockFn = jest.fn();

    const { getByTestId } = renderWithTheme(
      <ScrambledLetter
        showing={false}
        letter="S"
        isShuffling={false}
        disabled={false}
        scaleValue={new ReactNative.Animated.Value(1)}
        onPressLetter={mockFn}
      />,
    );

    const element = getByTestId("empty-placeholder");
    expect(element).not.toBeNull();
  });
});
