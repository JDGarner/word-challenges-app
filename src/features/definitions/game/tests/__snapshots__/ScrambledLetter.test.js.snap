// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrambled Letter should match snapshot 1`] = `
<View
  style={
    Object {
      "opacity": 1,
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
>
  <View
    style={
      Object {
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      accessible={true}
      className="sc-gqjmRU bsUHzu"
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={Array []}
    >
      <Text
        className="sc-ifAKCX lojWEY"
      >
        S
      </Text>
    </View>
  </View>
</View>
`;
